File: tool/kinetic_chain_viewer/set_colors.js
// Port of Hydrawave3/apps/web/src/anatomy/setColors.js — the per-set palette.
//
// Okabe-Ito: 8 visually distinct, colourblind-friendly hues.
//
// ORDER IS PRODUCT-SPECIFIED, not aesthetic: Set 1 Yellow, Set 2 Blue, Set 3
// Green, and the Sun-Moon connecting arc of each set takes that set's colour.
// `setIndex` is 0-BASED in every marker builder on both platforms, so index 0
// is Set 1. Sets 4+ are headroom — nothing in the corpus authors more than 3.
//
// The Flutter side must agree: `kAnatomySetColors` in the pad map screen mirrors
// this array, because "matching set colours" only means anything if the legend,
// the chips and the 3D badges all say the same thing.

export const SET_COLORS = [
  0xf0e442, // Set 1 — yellow
  0x0072b2, // Set 2 — blue
  0x009e73, // Set 3 — green
  0xd55e00, // Set 4 — vermillion
  0x56b4e9, // Set 5 — sky blue
  0xcc79a7, // Set 6 — reddish purple
  0xe69f00, // Set 7 — orange
  0x999999, // Set 8 — grey
];

export function perfSetColor(setIndex) {
  const n = SET_COLORS.length;
  return SET_COLORS[(((Number(setIndex) || 0) % n) + n) % n];
}

export function perfSetColorCss(setIndex) {
  return `#${perfSetColor(setIndex).toString(16).padStart(6, "0")}`;
}

File: lib/features/pad_placement/domain/set_colors.dart
import 'dart:ui' show Color;

/// Per-set palette, the Flutter port of Hydrawave3
/// `apps/web/src/anatomy/setColors.js`.
///
/// Kept in its own tiny module for the same reason the web does — no widgets, no
/// theme, no 3D — so the set list, the legend, the chips and the AnatomyScene
/// stage can all import it and cannot drift apart. "Matching set colours" only
/// means something if every surface agrees.
///
/// Okabe-Ito: 8 visually distinct, colourblind-friendly hues.
///
/// ORDER IS PRODUCT-SPECIFIED, not aesthetic — the Jul 25 CEO review asked for
/// Set 1 yellow, Set 2 blue, Set 3 green, and for each set's Sun→Moon connecting
/// line to take that set's colour. `setIndex` is 0-BASED in both marker builders,
/// so index 0 is Set 1. Sets 4+ keep the remaining Okabe-Ito hues; nothing in the
/// corpus authors more than 3 sets today, so those are headroom.
///
/// Deliberately NOT `RefPalette.set1/2/3`, which is a theme token used elsewhere
/// and has no reason to follow the 3D viewer.
const List<Color> kAnatomySetColors = [
  Color(0xFFF0E442), // Set 1 — yellow
  Color(0xFF0072B2), // Set 2 — blue
  Color(0xFF009E73), // Set 3 — green
  Color(0xFFD55E00), // Set 4 — vermillion
  Color(0xFF56B4E9), // Set 5 — sky blue
  Color(0xFFCC79A7), // Set 6 — reddish purple
  Color(0xFFE69F00), // Set 7 — orange
  Color(0xFF999999), // Set 8 — grey
];

/// The colour for a 0-based set index, wrapping like the web's `perfSetColor`.
Color anatomySetColor(int setIndex) {
  final n = kAnatomySetColors.length;
  return kAnatomySetColors[((setIndex % n) + n) % n];
}
